scipy/special/special/tools.h

```diff
@@ -73,5 +105,165 @@ namespace detail {
         return result;
     }
 
+    /* Performs one step of Kahan summation. */
+    template <typename T>
+    SPECFUN_HOST_DEVICE void kahan_step(T *sum, T *comp, T x) {
+        T y = x - *comp;
+        T t = *sum + y;
+        *comp = (t - *sum) - y;
+        *sum = t;
+    }
```

Comment:
I experimented a few ways to code Kahan summation, such as making a `KahanNumber` class that overloads `+=`, implementing a binary function that looks like `std::plus`, etc. All of these boil down to the computation of what the `kahan_step` function does, so I added that function. I ended up writing `series_eval_kahan` because I find it the most readable.

The benefit of having `kahan_step` as a separate function rather than “inlining” it is (1) for readability, and (2) to make it easier to verify code correctness.

I had a use case for `kahan_step` outside of `series_eval_kahan`. There are different possibilities of choosing `b_0` for the `iv_ratio` continued fraction. I compared three choices: $0$, $2\nu$, and $x+2\nu$. The middle one leads to the lowest RMSE (about 25% lower than the rest), which is intriguing because the first one has a slightly lower theoretical truncation error bound. This is partially resolved by using Kahan summation on the “outside”, though the middle one is still slightly better. So I keep the middle one in the final code.

In summary, I think it’s helpful to keep `kahan_step` as a stand-alone function.

Reply:
I see. I personally find it more readable to not have it be a separate function. It's a small enough number of lines that I can mentally chunk it as one thing, and then I find it easier to have everything in one place. Let's just keep it a separate function for now though. Later on, it might make more sense to have `series_eval` take a callable for the summation step which would default to a regular summation, but we could pass in a `kahan_step` or possibly other compensation methods.

Reply:
Ok!